### ISSUE TITLE
Clarify lifecycle of Constraint#predicate Predicate

### ISSUE THREAD

**Issue by findepi** [trinodb/trino#7620]
`Constraint#predicate` is an opaque `Predicate` that internally holds to an evaluator.
this is passed to `ConnectorMetadata#applyFilter`

Question:

- is connector allowed to hold onto this predicate after `ConnectorMetadata#applyFilter` returns?

Regardless of the answer, since this is a subtle aspect of the SPI, we should document that in the javadoc.

**Comment by martint** [comment]
No, it's not allowed to hold on to it. It's meant to be used only for the duration of the API call.

**Comment by findepi** [comment]
That would be my intuition as well, but i didn't know the intentions.

1. we should document
2. we can optionally enforce that, by "destroying" the evaluator after the call returns
